service/src/main/java/uk/nhs/adaptors/gpc/consumer/filters/SdsFilter.java

```diff
@@ -0,0 +1,162 @@
+package uk.nhs.adaptors.gpc.consumer.filters;
+
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.function.Function;
+
+import javax.annotation.PostConstruct;
+
+import org.apache.commons.lang3.StringUtils;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.cloud.gateway.filter.GatewayFilterChain;
+import org.springframework.cloud.gateway.filter.GlobalFilter;
+import org.springframework.cloud.gateway.filter.RouteToRequestUrlFilter;
+import org.springframework.cloud.gateway.support.ServerWebExchangeUtils;
+import org.springframework.core.Ordered;
+import org.springframework.http.HttpHeaders;
+import org.springframework.http.server.RequestPath;
+import org.springframework.http.server.reactive.ServerHttpRequest;
+import org.springframework.stereotype.Component;
+import org.springframework.util.CollectionUtils;
+import org.springframework.web.server.ServerWebExchange;
+
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import reactor.core.publisher.Mono;
+import uk.nhs.adaptors.gpc.consumer.sds.SdsClient;
+
+@Component
+@Slf4j
+@RequiredArgsConstructor(onConstructor = @__(@Autowired))
+public class SdsFilter implements GlobalFilter, Ordered {
+    private static final String GPC_URL_ENVIRONMENT_VARIABLE = "GP2GP_GPC_GET_URL";
+    private static final String INTERACTION_ID_PREFIX = "urn:nhs:names:services:gpconnect:";
+    private static final String STRUCTURED_ID = INTERACTION_ID_PREFIX + "fhir:operation:gpc.getstructuredrecord-1";
+    private static final String PATIENT_SEARCH_ID = INTERACTION_ID_PREFIX + "documents:fhir:rest:search:patient-1";
+    private static final String DOCUMENT_SEARCH_ID = INTERACTION_ID_PREFIX + "documents:fhir:rest:search:documentreference-1";
+    private static final String BINARY_READ_ID = INTERACTION_ID_PREFIX + "documents:fhir:rest:read:binary-1";
+    private static final String SSP_INTERACTION_ID = "Ssp-InteractionID";
+    private static final String SLASH = "/";
+    private static final String FHIR_SEPARATOR = "fhir/";
+    private static final String DOCUMENTS_SEPARATOR = "documents/";
+    private static final String DOUBLE_SEPARATOR = DOCUMENTS_SEPARATOR + FHIR_SEPARATOR;
+
+    private final SdsClient sdsClient;
+
+    private Map<String, Function<String, Optional<SdsClient.SdsResponseData>>> sdsRequestFunctions;
+
+    @Override
+    public Mono<Void> filter(ServerWebExchange exchange, GatewayFilterChain chain) {
+        if (StringUtils.isBlank(System.getProperty(GPC_URL_ENVIRONMENT_VARIABLE))) {
+            ServerHttpRequest serverHttpRequest = exchange.getRequest();
+            extractInteractionId(serverHttpRequest.getHeaders())
+                .ifPresent(id -> proceedSdsLookup(serverHttpRequest, exchange, id));
+        }
```

Comment:
Related to the comment above, we need to actually use the value of GPC_URL_ENVIRONMENT_VARIABLE.
```
} else {
   // use the value of GPC_URL_ENVIRONMENT_VARIABLE as the "endpoint" instead of using SDS to look it up
}
```

If GPCC_OVERRIDE_GPC_PROVIDER_ENDPOINT environment variable is set and non-empty then:
1)	The SDS API request should not occur
2)	And the value of this variable is used for [provider_service_root_url] for all interactions

Reply:
This is handled by the "uri" configuration in application.yml 👍

Reply:
If it is already in config then there is no need to set it again in the filter, right? It is being used by default instead of the sds one.